### PR TITLE
Add support for a CDN in front of crates/readmes

### PR DESCRIPTION
**Pull request by alexcrichton**
This should hopefully allow us to configure a CDN on Heroku (namely CloudFront)
to distribute crates/readmes through.

### CODE REVIEW

File: src/uploaders.rs [hunk 1]
```diff
@@ -61,11 +65,14 @@ impl Uploader {
     /// It returns `None` if the current `Uploader` is `NoOp`.
     pub fn readme_location(&self, crate_name: &str, version: &str) -> Option<String> {
         match *self {
-            Uploader::S3 { ref bucket, .. } => Some(format!(
-                "https://{}/{}",
-                bucket.host(),
-                Uploader::readme_path(crate_name, version)
-            )),
+            Uploader::S3 { ref bucket, .. } => {
```
Comment (carols10cents) [inline src/uploaders.rs]: Looks like this line needs to be changed to `Uploader::S3 { ref bucket, ref cdn, .. } => {`

**Comment by carols10cents** [comment]
@withoutboats is CloudFront reachable when using Tor?

**Comment by alexcrichton** [comment]
Oops forgot to verify last minute change, should be fixed now. It's also worth mentioning that the motivation for this is primarily https://github.com/rust-lang/cargo/issues/4670 where I was reminded that S3 is unreachable in China. The reporter there though mentioned that rust-lang.org is reachable (which uses CloudFront).

I've also historically wanted to use CloudFront instead of S3 directly for download speed and reliability, but nothing has been pressing except for the China-blocks-us-west-1 problem

**Comment by withoutboats** [comment]
I don't know anything about cloudfront but I know other CDNs (e.g. cloudflare) will frequently throw up captchas for tor users as a ddos mitigation. We ought to make sure the CDN will never do it in our use case because I expect cargo would just barf.

**Comment by alexcrichton** [comment]
Does www.rust-lang.org work? If so it's the same CDN. I don't know if CloudFront does captchas...

**Comment by withoutboats** [comment]
Based on this white paper I think cloudfront on its own will never intercept requests as long as we don't use a service like AWS WAF https://d0.awsstatic.com/whitepapers/DDoS_White_Paper_June2015.pdf

**Comment by carols10cents** [comment]
bors: r+

**Comment by messense** [comment]
seems unable to load readme now?

```
Refused to connect to 'https://d19xqa3lc3clo8.cloudfront.net/readmes/cc/cc-1.0.3.html' because it violates the following Content Security Policy directive: "connect-src 'self' https://docs.rs https://crates-io.s3-us-west-1.amazonaws.com".
```

**Comment by carols10cents** [comment]
@messense on it! Thank you for letting me know!

**Comment by carols10cents** [comment]
@messense readmes are fixed now!

**Pull request merged**